index.src.html

```diff
@@ -2394,14 +2420,692 @@
+    7. Fire a simple event named {{ImageTrackList/change}} at the
+        {{ImageDecoder/tracks}} object.
+
+: <dfn for=ImageDecoder>Decode Complete Frame</dfn> (with |frameIndex| and
+    |promise|)
+:: 1. Assert that {{ImageDecoder/[[tracks established]]}} is `true`.
+    2. Assert that {{ImageDecoder/[[internal selected track index]]}} is not
+        `-1`.
+    3. Let |encodedFrame| be the encoded frame identified by |frameIndex| and
+        {{ImageDecoder/[[internal selected track index]]}}.
+    4. Wait for any of the following conditions to be true (whichever happens
+        first):
+        1. {{ImageDecoder/[[encoded data]]}} contains enough bytes to
+            completely decode |encodedFrame|.
+        2. {{ImageDecoder/complete}} is `true`.
+        3. {{ImageDecoder/[[fatal error]]}} is `true`.
+    5. If {{ImageDecoder/[[encoded data]]}} does not contain enough bytes to
+        completely decode |encodedFrame|:
+        1. Remove |promise| from {{ImageDecoder/[[pending decode promises]]}}.
+        2. Reject |promise| with {{InvalidStateError}} {{DOMException}}.
+        3. Abort these steps.
+    6. Attempt to use {{ImageDecoder/[[codec implementation]]}} to decode
+        |encodedFrame|.
+    7. If decoding produces an error:
+        1. Remove |promise| from {{ImageDecoder/[[pending decode promises]]}}.
+        2. Reject |promise| with {{EncodingError}} {{DOMException}}.
+        3. Abort these steps.
+    8. If {{ImageDecoder/[[progressive frame generations]]}} contains an entry
+        keyed by |frameIndex|, remove the entry from the map.
+    9. Let |output| be the decoded image data emitted by
+        {{ImageDecoder/[[codec implementation]]}} corresponding to
+        |encodedFrame|.
+    10. Let |decodeResult| be a new {{ImageDecodeResult}} initialized as
+        follows:
+        1. Assign 'true' to {{ImageDecodeResult/complete}}.
+        2. Assign {{ImageDecodeResult/image}} with the result of running the
+            [=Create a VideoFrame=] algorithm with |output|.
+    13. Remove |promise| from {{ImageDecoder/[[pending decode promises]]}}.
+    14. Resolve |promise| with |decodeResult|.
+
+: <dfn for=ImageDecoder>Decode Progressive Frame</dfn> (with |frameIndex| and
+    |promise|)
+:: 1. Assert that {{ImageDecoder/[[tracks established]]}} is `true`.
+    2. Assert that {{ImageDecoder/[[internal selected track index]]}} is not
+        `-1`.
+    3. Let |encodedFrame| be the encoded frame identified by |frameIndex| and
+        {{ImageDecoder/[[internal selected track index]]}}.
+    4. Let |lastFrameGeneration| be `null`.
+    5. If {{ImageDecoder/[[progressive frame generations]]}} contains a map
+        entry with the key |frameIndex|, assign the value of the map entry to
+        |lastFrameGeneration|.
+    6. Wait for any of the following conditions to be true (whichever happens
+        first):
+        1. {{ImageDecoder/[[encoded data]]}} contains enough bytes to decode
```

Comment:
The current implementation just waits until _any_ new bytes have been received and retries to see if a new image can be decoded. Not sure if that's easier to phrase here.